server/src/main/java/io/crate/execution/dml/upsert/UpdateToInsert.java

```diff
@@ -0,0 +1,267 @@
+ *
+ *  Result:
+ *      targetColumns [x,            o]
+ *      values:       [1, {y=40, z=20}]
+ *  </pre>
+ *
+ *
+ *  <pre>
+ *   Table with columns (x, y, z) and existing record: (1, 2, 3)
+ *
+ *   INSERT INTO tbl (x, y, z) values (1, 10, 20)
+ *   ON CONFLICT (x) DO UPDATE SET
+ *      y = y + excluded.y
+ *
+ *  Result:
+ *      targetColumns: [x,  y, z]
+ *      values:        [1, 12, 3]
+ **/
+public final class UpdateToInsert {
+
+    private final DocTableInfo table;
+    private final Evaluator eval;
+    private final List<Reference> updateColumns;
+    private final ArrayList<Reference> columns;
+
+    record Values(Doc doc, Object[] excludedValues) {
+    }
+
+    private static class Evaluator extends BaseImplementationSymbolVisitor<Values> {
+
+        private final ReferenceResolver<CollectExpression<Doc, ?>> refResolver;
+
+        private Evaluator(NodeContext nodeCtx,
+                          TransactionContext txnCtx,
+                          ReferenceResolver<CollectExpression<Doc, ?>> refResolver) {
+            super(txnCtx, nodeCtx);
+            this.refResolver = refResolver;
+        }
+
+        @Override
+        public Input<?> visitInputColumn(InputColumn inputColumn, Values context) {
+            return Literal.ofUnchecked(inputColumn.valueType(), context.excludedValues[inputColumn.index()]);
+        }
+
+        @Override
+        public Input<?> visitReference(Reference symbol, Values values) {
+            CollectExpression<Doc, ?> expr = refResolver.getImplementation(symbol);
+            expr.setNextRow(values.doc);
+            return expr;
+        }
+    }
+
+    static class RefResolver implements ReferenceResolver<CollectExpression<Object[], Object>> {
+
+        private final Collection<Reference> columns;
+
+        public RefResolver(Collection<Reference> columns) {
+            this.columns = columns;
+        }
+
+        @Override
+        public CollectExpression<Object[], Object> getImplementation(Reference ref) {
+            int idx = Reference.indexOf(columns, ref.column());
+            assert idx > -1 : "Column " + ref + " must exist in " + columns;
+            return NestableCollectExpression.forFunction(cells -> cells[idx]);
+        }
+    }
+
+    public UpdateToInsert(NodeContext nodeCtx,
+                          TransactionContext txnCtx,
+                          DocTableInfo table,
+                          String[] updateColumns) {
+        var refResolver = new DocRefResolver(table.partitionedBy());
+        this.table = table;
+        this.eval = new Evaluator(nodeCtx, txnCtx, refResolver);
+        this.updateColumns = new ArrayList<>(updateColumns.length);
+        this.columns = new ArrayList<>();
+        boolean errorOnUnknownObjectKey = txnCtx.sessionSettings().errorOnUnknownObjectKey();
+        for (var column : table.columns()) {
+            if (column instanceof GeneratedReference && !Arrays.asList(updateColumns).contains(column.column().fqn())) {
```

Comment:
I don't understand the second part, if it's explicitly contained, do we provide values for a generated column?

Reply:
Added a comment